pageserver/src/tenant/timeline.rs

```diff
@@ -4176,6 +4177,398 @@ impl Timeline {
+    /// - has prev_lsn in remote storage (temporary restriction)
+    ///
+    /// After the operation has been started, it cannot be canceled. Upon restart it needs to be
+    /// polled again until completion.
+    ///
+    /// During the operation all timelines sharing the data with this timeline will be reparented
+    /// from our ancestor to be branches of this timeline.
+    pub(crate) async fn prepare_to_detach_from_ancestor(
+        self: &Arc<Timeline>,
+        tenant: &crate::tenant::Tenant,
+        options: AncestorDetachOptions,
+        ctx: &RequestContext,
+    ) -> Result<PreparedTimelineDetach, DetachFromAncestorError> {
+        use DetachFromAncestorError::*;
+
+        if self.remote_client.as_ref().is_none() {
+            unimplemented!("no new code for running without remote storage");
+        }
+
+        let Some((ancestor, ancestor_lsn)) = self
+            .ancestor_timeline
+            .as_ref()
+            .map(|tl| (tl.clone(), self.ancestor_lsn))
+        else {
+            return Err(NoAncestor);
+        };
+
+        if !ancestor_lsn.is_valid() {
+            return Err(NoAncestor);
+        }
+
+        if ancestor.ancestor_timeline.is_some() {
+            // non-technical requirement; we could flatten N ancestors just as easily but we chose
+            // not to
+            return Err(TooManyAncestors);
+        }
+
+        if self.get_prev_record_lsn() == Lsn::INVALID
+            || self.disk_consistent_lsn.load() == ancestor_lsn
+        {
+            // this is to avoid a problem that after detaching we would be unable to start up the
+            // compute because of "PREV_LSN: invalid".
+            return Err(DetachedTimelineNeedsWrites);
+        }
+
+        // before we acquire the gate, we must mark the ancestor as having a detach operation
+        // ongoing which will block other concurrent detach operations so we don't get to ackward
+        // situations where there would be two branches trying to reparent earlier branches.
+        let (_guard, barrier) = utils::completion::channel();
+
+        {
+            let mut guard = tenant.ongoing_timeline_detach.lock().unwrap();
+            if let Some((tl, other)) = guard.as_ref() {
+                if !other.is_ready() {
+                    return Err(OtherTimelineDetachOngoing(*tl));
+                }
+            }
+            *guard = Some((self.timeline_id, barrier));
+        }
+
+        let _gate_entered = self.gate.enter().map_err(|_| ShuttingDown)?;
+
+        if ancestor_lsn >= ancestor.get_disk_consistent_lsn() {
+            ancestor
+                .freeze_and_flush()
+                .await
+                .map_err(DetachFromAncestorError::FlushAncestor)?;
+
+            // we do not need to wait for uploads to complete, because we only want to see the file
+            // on disk.
+            tracing::info!("froze and flush the ancestor");
+        }
+
+        let end_lsn = ancestor_lsn + 1;
```

Comment:
Why is the code past this line not part of the `detach_ancestors` module?
I feel like that's where it belongs.

Also, I think there are two subroutines in the code below, that would benefit from being in separate functions.

One subroutine for the part where we do the rewriting.
Another subroutine for the part where we only need to do the remote copy + local hardlinking.

Reply:
Yeah, it would be nice to move both prepare and completion to the detach_ancestor.rs. I will just need to rewrite it from `&self` which I will not enjoy, but I'll do it as last.